<a href="https://colab.research.google.com/github/priyaregit/EDA_Hotel-Booking/blob/main/Hotel_Booking_Analysis_EDA_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests? This hotel booking dataset can help you explore those questions!

## <b>This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. All personally identifying information has been removed from the data. </b>

## <b> Explore and analyze the data to discover important factors that govern the bookings. </b>

In [2]:
# Loading Google Drive

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
# Importing Pandas and Numpy libraries for the purpose of data analysis

import pandas as pd
import numpy as np


# Importing Matplotlib and Seaborn libraries for data visualization

import matplotlib.pyplot as plt
import seaborn as sns

### Let us first read the dataset from Google drive

In [116]:
# Reading the dataset from Google drive and storing it in the dataframe: hotel_booking

hotel_booking = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Capstone Projects/EDA/Data/Hotel Bookings.csv')

### We must perform some basic checks on the dataset such as:

* ### Checking the shape of dataframe i.e. number of rows and columns

* ### Checking for the various variables/features of our dataset

* ### Checking the data type of each of the features

* ### Checking for the missing values/ null values

### These checks would give us a baseline understanding of the available data

In [117]:
# Having a look at the first 5 rows/observations from our data

hotel_booking.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

In [118]:
# Checking the last 3 rows from the data

hotel_booking.tail(3)   

hotel  is_canceled  lead_time  arrival_date_year  \
119387  City Hotel            0         34               2017   
119388  City Hotel            0        109               2017   
119389  City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
119387                            0                          4   
119388                            0                          0   
119389                            0                          2   

        reservation_status reservation_status_date  
119387           Check-Out              2017-09-07  
119388           Check-Out              2017-09-07  
119389           Check-Out              2017-09-07  

[3 rows x 32 columns]

In [119]:
# Checking the number of rows and columns in the dataset

hotel_booking.shape

(119390, 32)

### There are 119,390 rows and 32 columns in the dataset

In [120]:
# Checking for information such as missing values and data type of all the columns 

hotel_booking.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

### We find that there are 4 columns- children, country, agent and company- with missing/null values. We will deal with the issue of missing values a little later.

### We also find that our data is of object, integer and float data types.

### There are 20 columns with numeric data (integer and float data type) and 12 columns with categorical data. 

### However, on closer observation, we notice that the column 'reservation_status_date' which contains dates, is of object data type. We need to change its data type


In [121]:
# Importing datetime module 

from datetime import datetime

# Converting data type from object to datetime
hotel_booking['reservation_status_date'] =  pd.to_datetime(hotel_booking['reservation_status_date'], format="%Y-%m-%d")

### Let us now begin our data exploration by creating a Data Dictionary

### First, we need to find the unique values present in all the columns containing categorical data.

In [122]:
# Creating a list of all the column names

column_names = list(hotel_booking.columns)


# Running a 'for' loop across all the 32 columns and printing the 'unique' values 'if' the column is of object data type

for i in range(len(hotel_booking.columns)):                  

  if hotel_booking.dtypes[column_names[i]] == 'object':    

    print(f"The unique values for {column_names[i]} are: {list(hotel_booking[column_names[i]].unique())}") # 


The unique values for hotel are: ['Resort Hotel', 'City Hotel']
The unique values for arrival_date_month are: ['July', 'August', 'September', 'October', 'November', 'December', 'January', 'February', 'March', 'April', 'May', 'June']
The unique values for meal are: ['BB', 'FB', 'HB', 'SC', 'Undefined']
The unique values for country are: ['PRT', 'GBR', 'USA', 'ESP', 'IRL', 'FRA', nan, 'ROU', 'NOR', 'OMN', 'ARG', 'POL', 'DEU', 'BEL', 'CHE', 'CN', 'GRC', 'ITA', 'NLD', 'DNK', 'RUS', 'SWE', 'AUS', 'EST', 'CZE', 'BRA', 'FIN', 'MOZ', 'BWA', 'LUX', 'SVN', 'ALB', 'IND', 'CHN', 'MEX', 'MAR', 'UKR', 'SMR', 'LVA', 'PRI', 'SRB', 'CHL', 'AUT', 'BLR', 'LTU', 'TUR', 'ZAF', 'AGO', 'ISR', 'CYM', 'ZMB', 'CPV', 'ZWE', 'DZA', 'KOR', 'CRI', 'HUN', 'ARE', 'TUN', 'JAM', 'HRV', 'HKG', 'IRN', 'GEO', 'AND', 'GIB', 'URY', 'JEY', 'CAF', 'CYP', 'COL', 'GGY', 'KWT', 'NGA', 'MDV', 'VEN', 'SVK', 'FJI', 'KAZ', 'PAK', 'IDN', 'LBN', 'PHL', 'SEN', 'SYC', 'AZE', 'BHR', 'NZL', 'THA', 'DOM', 'MKD', 'MYS', 'ARM', 'JPN', 'LKA',

### Now that we know the unique labels in each of the categorical columns, let us understand the information contained in all the columns through a Data Dictionary.

###  The various columns are as follows:

### 1 hotel: It provides information about the type of hotel booked. There are two types of hotels- Resort Hotel and City Hotel

### 2 is_canceled: It indicates if the booking was cancelled. It assigns the value of 1 if the booking is cancelled, else it is assigned 0

### 3 lead_time: It gives the number of days that elapsed between the date of  booking and the arrival date of the guests

### 4 arrival_date_year: It gives the year of the guest's arrival date. We have data for the years 2015, 2016 and 2017

### 5 arrival_date_month: It is a categorical value that gives the month of the guest's arrival date with 12 categories- “January” to “December”

### 6 arrival_date_week_number:	It gives the number of the week of the guest's arrival date. It could range from 1 to 53

### 7 arrival_date_day_of_month: It gives the day of the month on which the guest's arrive. It could range from 1 to 31

### 8 stays_in_weekend_nights: It indicates the number of weekend nights (Saturday or Sunday) that the guests stayed or booked to stay at the hotel

### 9 stays_in_week_nights: It indicates the number of week nights (Monday to Friday) that the guests stayed or booked to stay at the hotel

### 10 adults: It shows the number of adults who were booked for a stay at the hotel

### 11 children: It shows the number of children (age group not indicated) who were booked for a stay at the hotel

### 12 babies: It shows the number of babies (age group not indicated) who were booked for a stay at the hotel

### 13 meal: It is a categorical variable that indicates the type of meal booked. Categories present (in accordance with standard hospitality meal packages) are:

 * ### BB – Bed & Breakfast
 * ### HB – Half Board (breakfast and one other meal)
 * ### FB – Full Board (breakfast, lunch and dinner)
 * ### SC/Undefined – no meal package

### 14 country: It indicates the guest's country of origin

### 15 market_segment: It is the designation of the market segment through which the booking was made. The various categories are: Direct, Corporate, Online TA, Offline TA/TO, Complementary, Groups, Aviation and Undefined. Here, the term 'TA' means Travel Agents and 'TO' means Tour Operators

### 16 distribution_channel: It is a categorical	variable indicating the distribution channel through which the booking was made. The various categories are: Direct, Corporate, TA/TO, GDS (Global Distribution System) and Undefined

### 17 is_repeated_guest: It indicates if the booking was made by a repeated guest (1) or not (0)

### 18 previous_cancellations: It gives the number of previous bookings that were cancelled by the customer prior to the current booking

### 19 previous_bookings_not_canceled: It gives the number of previous bookings that were not cancelled by the customer prior to the current booking

### 20 reserved_room_type: It is a categorical	code for the type of room that has been reserved while booking. It may be 'C', 'A', 'D', 'E', 'G', 'F', 'H', 'L', 'P' or 'B'

### 21 assigned_room_type: It is a categorical	code for the type of room assigned to the booking. It may be 'C', 'A', 'D', 'E', 'G', 'F', 'I', 'B', 'H', 'P', 'L' or 'K'

### 22 booking_changes: It indicates the number of changes made to the booking between the time of booking and the moment of check-in or cancellation

### 23 deposit_type: It gives a categorical indication of the type of deposit made by the customer to confirm the booking. This variable assumes three categories- 'No Deposit', 'Refundable' or 'Non Refund'

### 24 agent: It is a marker for the travel agency that has made the booking
      
### 25 company: It is a marker of the company/entity that has made the booking

### 26 days_in_waiting_list: It gives the number of days the booking was in the waiting list before it was confirmed to the customer

### 27 customer_type: It is a categorical variable that informs about the type of customer. It may be Contract, Group, Transient or Transient-Party

### 28 adr: It is the Average Daily Rate which is a measure of the average revenue that the hotel receives for each occupied guest room per day

### 29 required_car_parking_spaces: It gives the number of car parking spaces required by the customer

### 30 total_of_special_requests: It gives the number of special requests made by the customer (e.g. twin bed or high floor)

### 31 reservation_status: It is a categorical variable indicating the last known status of the reservation. It may be any of the three categories- Check-Out, Canceled or No-Show 

### 32 reservation_status_date: It gives the date on which the reservation_status was last set.		








### Having checked what type of information each of the columns provide, let us now inspect the numeric data

In [123]:
# Checking the key statistics of the all the columns with numeric data

hotel_booking.describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2016.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2015.000000   
25%         0.000000      18.000000        2016.000000   
50%         0.000000      69.000000        2016.000000   
75%         1.000000     160.000000        2017.000000   
max         1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119390.000000         119390.000000  119390.000000   
mean                  0.927599              2.500302       1.856403   
std                   0.998613              1.908286       0.579261   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes          agent      company  days_in_waiting_list  \
count    119390.000000  103050.000000  6797.000000         119390.000000   
mean          0.221124      86.693382   189.266735              2.321149   
std           0.652306     110.774548   131.655015             17.594721   
min           0.000000       1.000000     6.000000              0.000000   
25%           0.000000       9.000000    62.000000              0.000000   
50%           0.000000      14.000000   179.000000              0.000000   
75%           0.000000     229.000000   270.000000              0.000000   
max          21.000000     535.000000   543.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119390.000000                119390.000000              119390.000000  
mean      101.831122                     0.062518                   0.571363  
std        50.535790                     0.245291                   0.792798  
min        -6.380000                     0.000000                   0.000000  
25%        69.290000                     0.000000                   0.000000  
50%        94.57500

### Some inferences that we can draw from the above statistics are:

### * The average lead time is 104 days with the maximum being 737 days.

### * There could be outliers in columns such as 'previous_cancellations',	'previous_bookings_not_canceled' and 'booking_changes'

### * The mean 'adr' is 101.83

### * The mean of 'days_in_waiting_list' is 2.32 days. However, the maximum waitlist period is 391 days

### We would get a clearer picture of the hotel booking data through visualisation of all the individual statistics.

### But before we proceed any further, there is an important task to accomplish- the detection of missing/null values and dealing with them.


## <b><u> Handling Missing Values 


### We have learnt earlier that there are 4 columns with missing values. Let us gather some more information about them


In [124]:
# Creating a function to find the missing values in our dataset

def missing():

# Finding the number of missing/null values across all columns in the dataset
  missing_values = hotel_booking.isnull().sum()

# Finding the count of missing values in only those columns which have them
  mis_val_count = missing_values[missing_values > 0]

# Finding the % of missing values in the columns
  mis_val_percent = round(100 * mis_val_count / len(hotel_booking), 2)

# Creating a table with the above information and renaming the column heads
  mis_value_table = pd.concat([mis_val_count, mis_val_percent], axis=1)
  mis_value_table = mis_value_table.rename(columns = {0 : 'Missing Values', 1 : '% Missing'})

# If there are missing values, print the mis_value_table. Else, print that there are no missing values
  if mis_value_table.shape[0] != 0:
    missing_result = mis_value_table
  else:
     missing_result = "There are no missing/null values in the hotel_booking dataset"

  return missing_result

In [125]:
# Calling the function for missing values

missing()

Missing Values  % Missing
children               4       0.00
country              488       0.41
agent              16340      13.69
company           112593      94.31

### There are 4 observations in the 'children' column that have missing/null values. 
### We can deal with it in two ways. 
###  * Since they are only four (i.e. insignificant in our dataset) in number, we can remove the rows/observations with missing values for 'children' column. 
###  * Alternatively, we can assign 0 to these observations as the mean value in that column (0.103890) is also nearly 0
### Here, I choose to fill the missing/null observations in the 'children' column with 0



In [126]:
# Filling the missing/null observations in the 'children' column with 0

hotel_booking['children'].fillna(0,inplace = True)

### There are 488 observations in the 'country' column that have missing/null values. 
### We have three options available with us. 
###  * Since values are missing in only 0.41% of the total rows, we may choose to delete such rows. 
###  * 'country' being a categorical column, we may choose to replace the missing values with the most common/ frequently occurring class. 
###  * The third alternative is to simply, replace the missing values with the term "Unknown". 
### Here, I would like to go with the third option. Exercising this option would not influence our dataset in any undue way and will also keep the original message (that the country of origin for these observations is unknown) intact.


In [127]:
# Replacing the missing/null observations in the 'country' column with 'Unknown'

hotel_booking['country'].fillna('Unknown',inplace = True)

### There are 16340	observations in the 'agent' column that have missing/null values. This translates to 13.69% of the total observations.
### The missing/null values constitute more than 10% of the observations. So, it would not be appropriate to remove those rows/observations as there will be a significant loss of data 
###  The 'agent' column is a marker for the unique IDs of the travel agency that has made the bookings. So, we cannot proceed with the imputation of mean value
###  Again, we have two options ahead of us:
### * We may choose to impute the mode i.e. populate the missing values with the ID of the agent with most number of bookings
### * We could fill all the missing observations with the value 0 as a marker for the unknown agent

In [128]:
# Finding the number of unique agents who made the bookings

print(f"The number of unique 'agent' are {hotel_booking.agent.nunique()}","\n")

# Finding the number of bookings made by each of the unique agents 

print("The number of bookings made by each 'agent' are: ")
booking_by_agent = hotel_booking['agent'].value_counts().reset_index()
booking_by_agent.rename(columns = {'index' : 'Agent', 'agent' : 'Booking_count'},inplace=True)
booking_by_agent

The number of unique 'agent' are 333 

The number of bookings made by each 'agent' are: 


Agent  Booking_count
0      9.0          31961
1    240.0          13922
2      1.0           7191
3     14.0           3640
4      7.0           3539
..     ...            ...
328  289.0              1
329  432.0              1
330  265.0              1
331   93.0              1
332  304.0              1

[333 rows x 2 columns]

### We see that there are 333 unique agents who have made bookings. However, the distribution of bookings made by agents seems to be very skewed. Maximum number of bookings (31,961) were made by agent marked '9'. The second highest number of bookings (13,922) made by agent marked '240' is about 43% lesser than the highest number. 
### By imputing mode of the 'agent' data to the missing values (16340 in number), the already skewed distribution will become even more skewed. Hence, I would fill all the missing observations with the value 0 as a marker for bookings by an unknown/no agent


In [129]:
# Filling the missing/null observations in the 'country' column with 0

hotel_booking['agent'].fillna(0,inplace = True)

### We are now left with only the 'company' column with missing/null values in it. 
### A whopping 112,593	rows or 94.31% of the total datset have missing values for the 'company' feature. With so much missing information, we can ignore this column altogether.
### Therefore, i choose to drop the column 'company' from the dataset

In [131]:
# Dropping the column 'company' from hotel_booking dataset

hotel_booking.drop('company', axis=1, inplace=True)

### Let us quickly reconfirm the dropping of a column/feature and handling of missing values.

In [133]:
# Checking the shape of the dataset

hotel_booking.shape

(119390, 31)

In [132]:
# Checking for missing/null values across all columns

missing()

'There are no missing/null values in the hotel_booking dataset'